### PR TITLE
Fixing direction cosines

### PR DESCRIPTION
**Pull request by zjmorgan**
**Description of work.**

The algorithms `SaveHKL` and `SaveHKLCW` incorrectly assumes the the unit vector directions of the scattered and reverse incident beams are the direction cosines. I reality, to make these work for `Jana2020` or `WinGX`, the actual direction cosines need to be computed by performing one more step by taking the dot product with the crystallographic directions. This is clearly spelled out in the reference `A. Katrusiak, Zeitschrift für Kristallographie - Crystalline Materials 216, 646 (2001).`

Historically, this has worked with internal TOPAZ reduction scripts that computes it correctly. The unit tests are updated with a TOPAZ example. This has been tested to work with the extinction corrections in `Jana2020` and the absorption correction program in `WinGX`.

Added some updates to the user docs.

image

**To test:**

Run the unit tests to make sure they pass. Optionally test with  `Jana2020` and `WinGX`. They won't work if the direction cosines are wrong.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Crystal/src/SaveHKL.cpp [hunk 1]
```diff
@@ -441,11 +441,23 @@ void SaveHKL::exec() {
...
+          Kernel::DblMatrix R = p.getGoniometerMatrix();
 
-          auto U = ol.getU();
-          auto RU = oriented * U;
-          RU.Transpose();
+          R.Transpose();
+
+          DblMatrix T = ol.getUB();
+
+          V3D t1 = T * V3D(1, 0, 0);
```
Comment (RichardWaiteSTFC) [inline Framework/Crystal/src/SaveHKL.cpp:450]: Aren't `tn` just the nth column of T?  It doesn't look like there's a `getCol` method for a `DblMatrix` but one could always init the `V3D` by accessing the values directly?

File: Framework/Crystal/src/SaveHKL.cpp [hunk 2]
```diff
@@ -441,11 +441,23 @@ void SaveHKL::exec() {
...
+          V3D t2 = T * V3D(0, 1, 0);
+          V3D t3 = T * V3D(0, 0, 1);
+
+          t1.normalize();
+          t2.normalize();
+          t3.normalize();
+
+          T.setRow(0, t1);
+          T.setRow(1, t2);
+          T.setRow(2, t3);
```
Comment (RichardWaiteSTFC) [inline Framework/Crystal/src/SaveHKL.cpp:460]: Is this equivalent to transposing `T` (after normalising columns)? 
If one made a copy of the oriented lattice, then set `a = b = c = 1` would the transpose of `UB` then be the same as this matrix `T`?

**Review by RichardWaiteSTFC — COMMENTED** [review]
I haven't got the means to test this in Jana at the moment, but here are some comments - mostly just variable names and maybe a shortcut (although perhaps it's less readable/transparent?)

File: Framework/Crystal/src/SaveHKL.cpp [hunk 3]
```diff
@@ -454,8 +466,11 @@ void SaveHKL::exec() {
           V3D dir = p.getDetPos() - inst->getSample()->getPos();
           dir.normalize();
 
-          V3D dir_cos_1 = RU * reverse_incident;
-          V3D dir_cos_2 = RU * dir;
+          V3D up = R * reverse_incident;
+          V3D us = R * dir;
```
[the remark this answers is not in the document]
Comment (zjmorgan) [reply Framework/Crystal/src/SaveHKL.cpp]: Good suggestion. I changed the variable names to be clearer.

File: Framework/Crystal/src/SaveHKL.cpp [hunk 4]
```diff
@@ -454,8 +466,11 @@ void SaveHKL::exec() {
...
+
+          V3D dir_cos_1 = T * up;
```
[the remark this answers is not in the document]
Comment (zjmorgan) [reply Framework/Crystal/src/SaveHKL.cpp]: Good suggestion. I changed the variable names to be more clear.

File: Framework/Crystal/src/SaveHKL.cpp (hunk 1 above)
Comment (zjmorgan) [reply Framework/Crystal/src/SaveHKL.cpp]: They are the columns of `T` and you are right, there is no `getCol` method. I initialize the V3D explicitly now.

File: Framework/Crystal/src/SaveHKL.cpp (hunk 2 above)
Comment (zjmorgan) [reply Framework/Crystal/src/SaveHKL.cpp:460]: This is the transpose. Your second statement would be correct as it would be the identity.

**Comment by zjmorgan** [comment]
Also changed the `SaveHKLCW` tests to be more complicated comparing against the TOPAZ reduction software.

**Comment by RichardWaiteSTFC** [comment]
Thanks for the changes - they look good. I'm just wondering whether the calculation of the matrix `T` belongs in a method of the oriented lattice e.g. `getT()` or similar (along with some documentation about this matrix and the direction cosines). It's already used in two algorithms (`SaveHKL ` and `SaveHKLCW`) and may be required elsewhere (e.g. `SaveReflections`, algorithms doing some absorption corrections etc.).

**Comment by zjmorgan** [comment]
> Thanks for the changes - they look good. I'm just wondering whether the calculation of the matrix `T` belongs in a method of the oriented lattice e.g. `getT()` or similar (along with some documentation about this matrix and the direction cosines). It's already used in two algorithms (`SaveHKL ` and `SaveHKLCW`) and may be required elsewhere (e.g. `SaveReflections`, algorithms doing some absorption corrections etc.).

That could be possible. I think the worry is whether we defined `T` according to the `B`-matrix coordinate system or the `UB`-matrix coordinate system

**Comment by RichardWaiteSTFC** [comment]
> > Thanks for the changes - they look good. I'm just wondering whether the calculation of the matrix `T` belongs in a method of the oriented lattice e.g. `getT()` or similar (along with some documentation about this matrix and the direction cosines). It's already used in two algorithms (`SaveHKL ` and `SaveHKLCW`) and may be required elsewhere (e.g. `SaveReflections`, algorithms doing some absorption corrections etc.).
> 
> That could be possible. I think the worry is whether we defined `T` according to the `B`-matrix coordinate system or the `UB`-matrix coordinate system

So in both algorithms with your change we do it in the `UB` frame - I think this makes sense, I can't think of why anyone would want this in the `B` frame without the `U` rotation, can you? Anyway happy for this to be done separately to this PR if this needs to be merged for scientists.

**Comment by zjmorgan** [comment]
> > > Thanks for the changes - they look good. I'm just wondering whether the calculation of the matrix `T` belongs in a method of the oriented lattice e.g. `getT()` or similar (along with some documentation about this matrix and the direction cosines). It's already used in two algorithms (`SaveHKL ` and `SaveHKLCW`) and may be required elsewhere (e.g. `SaveReflections`, algorithms doing some absorption corrections etc.).
> > 
> > 
> > That could be possible. I think the worry is whether we defined `T` according to the `B`-matrix coordinate system or the `UB`-matrix coordinate system
> 
> So in both algorithms with your change we do it in the `UB` frame - I think this makes sense, I can't think of why anyone would want this in the `B` frame without the `U` rotation, can you? Anyway happy for this to be done separately to this PR if this needs to be merged for scientists.

I think we can add that in. Would you be okay if I just make the method `getDirectionCosine` and attach it to `orientedLattice`? That way we can just have one place to calculate the values. I don't think it belongs to any `PeakWorkspace` because you need to define a lattice to compute correctly.

**Comment by zjmorgan** [comment]
Something like 
```python
ol.getDirectionCosine(Qvec)
```

**Comment by zjmorgan** [comment]
> Something like
> 
> ```python
> ol.getDirectionCosine(Qvec)
> ```


Thinking about it more, that won't work. You need the original coordinates. Better to leave outside.

File: Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp [hunk 5]
```diff
@@ -165,6 +165,9 @@ void export_IPeak() {
       .def("setGoniometerMatrix", &setGoniometerMatrix, (arg("self"), arg("goniometerMatrix")),
            "Set the :class:`~mantid.geometry.Goniometer` rotation matrix of "
            "this peak.")
+      .def("getDetectorID", &IPeak::getDetectorID, arg("self"),
```
Comment (RichardWaiteSTFC) [inline Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp:168]: Thanks, I've been meaning to expose this to python for some time!
The doc string states that this works only for rectangular detectors, however peak tables in WISH do have the correct DetID populated, so perhaps it works in general?

**Review by RichardWaiteSTFC — COMMENTED** [review]
Looks great, I like that the `T` matrix calc is now done in the `OrientedLattice` and that there is a new method exposed to calculate `cosFromDir`. Thanks also for adding some methods to the `Peak`/`LeanPeak` objects to get ki and kf unit vectors in the sample frame. Code looks good and tests are very thorough! Just one comment about a doc string possibly being outdated - once addressed I'll approve!

File: Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp (hunk 5 above)
Comment (zjmorgan) [reply Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp:168]: It is not tied to rectangular detectors, I agree.

**Review by RichardWaiteSTFC — APPROVED** [review]
Great work thanks for this!

**Pull request merged**